### DISCUSSION TITLE
docs: viper and garbage collection

### DISCUSSION THREAD

**Discussion by massimosala** [micropython/micropython#15257] [General]
### Documentation URL

https://docs.micropython.org/en/v1.7/pyboard/reference/speed_python.html

### Description

Let say the garbage collector is enabled.

During the execution of functions decorated with *viper*, can the gc collector kick in or not?

Sorry I don't find this specification in the docs.

### Code of Conduct

Yes, I agree

**Comment by andrewleech** [comment]
In my understanding gc.collect kicks in automatically when you try to allocated ram for an object and that allocation fails due to lack of space - then gc.collect will be run and the allocation retried.

viper functions can certainly allocate object / types that take ram so when that's done I'm sure gc.collect can be run.

**Comment by peterhinch** [comment]
By the same token if a Viper function doesn't allocate then GC won't kick in.

I've converted this to a discussion as action from the maintainers is not required.

**Comment by massimosala** [comment]
@andrewleech, @peterhinch 

Thanks to you for the clarifications.

As a principle, I don't allocate new objects inside _viper_ functions (apart from _viper_ integers).

Unfortunately in a few cases these functions have to call normal Python functions... and who knows exactly their internals!
In such cases, I will disable and re-enable the gc collector.

**Reply by andrewleech** [reply]
I generally use viper in situations where I'm looping over a lot of data. Inside the loops it's definitely worth trying to avoid slower things like allocations, but I'll often create buffers or convert types in the viper function just outside of a loop.
It feels risky to disable gc altogether, though that just means that instead of an allocation you'll get a MemoryError exception thrown which you'll need to catch and handle.
